File: iosApp/iosApp/ViewModels/SettingsViewModel.swift
import Foundation
import SharedLogic

@MainActor
public final class SettingsViewModel: ObservableObject {
    public let availableThemes: [AppThemePreset] = AppThemePreset.entries
    public let availableLanguages: [AppLanguage] = AppLanguage.entries

    @Published public private(set) var activeSourceName: String = "StoryVerse Drive Catalog"

    private let themeRepository: ThemeRepository
    private let localizationRepository: LocalizationRepository
    private let storySourceRegistry: StorySourceRegistry

    public init(
        themeRepository: ThemeRepository = KoinHelper().themeRepository,
        localizationRepository: LocalizationRepository = KoinHelper().localizationRepository,
        storySourceRegistry: StorySourceRegistry = KoinHelper().storySourceRegistry
    ) {
        self.themeRepository = themeRepository
        self.localizationRepository = localizationRepository
        self.storySourceRegistry = storySourceRegistry

        if let defaultSource = storySourceRegistry.getDefaultSource() {
            self.activeSourceName = defaultSource.metadata.name
        }
    }

    public func selectTheme(_ preset: AppThemePreset) {
        ThemeManager.shared.selectPreset(preset)
    }

    public func selectLanguage(_ language: AppLanguage) {
        LocalizationManager.shared.selectLanguage(language)
    }

    public func getColors(for preset: AppThemePreset) -> StoryVerseColors {
        ThemeManager.shared.getColors(for: preset)
    }
}

File: androidApp/src/main/kotlin/com/slowbuild/storyverse/storyverse/ui/settings/SettingsViewModel.kt
package com.slowbuild.storyverse.storyverse.ui.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.slowbuild.storyverse.domain.i18n.AppLanguage
import com.slowbuild.storyverse.domain.i18n.LocalizationRepository
import com.slowbuild.storyverse.domain.source.StorySourceRegistry
import com.slowbuild.storyverse.domain.theme.AppThemePreset
import com.slowbuild.storyverse.domain.theme.ThemeColors
import com.slowbuild.storyverse.domain.theme.ThemeRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class SettingsUiState(
    val currentThemePreset: AppThemePreset = AppThemePreset.LIGHT,
    val currentLanguage: AppLanguage = AppLanguage.VIETNAMESE,
    val availableThemes: List<AppThemePreset> = AppThemePreset.entries,
    val availableLanguages: List<AppLanguage> = AppLanguage.entries,
    val activeSourceName: String = ""
)

class SettingsViewModel(
    private val themeRepository: ThemeRepository,
    private val localizationRepository: LocalizationRepository,
    private val storySourceRegistry: StorySourceRegistry
) : ViewModel() {

    private val _uiState = MutableStateFlow(SettingsUiState())
    val uiState: StateFlow<SettingsUiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            themeRepository.currentPreset.collect { preset ->
                _uiState.update { it.copy(currentThemePreset = preset) }
            }
        }

        viewModelScope.launch {
            localizationRepository.currentLanguage.collect { lang ->
                _uiState.update { it.copy(currentLanguage = lang) }
            }
        }

        val defaultSource = storySourceRegistry.getDefaultSource()
        _uiState.update {
            it.copy(
                activeSourceName = defaultSource?.metadata?.name ?: "StoryVerse Drive Catalog"
            )
        }
    }

    fun selectTheme(preset: AppThemePreset) {
        viewModelScope.launch {
            themeRepository.setPreset(preset)
        }
    }

    fun selectLanguage(language: AppLanguage) {
        viewModelScope.launch {
            localizationRepository.setLanguage(language)
        }
    }

    fun getColorsForPreset(preset: AppThemePreset): ThemeColors {
        return themeRepository.getColorsForPreset(preset)
    }
}
